src/test/util/net.h

```diff
@@ -198,4 +201,105 @@ class StaticContentsSock : public ZeroSock
     mutable size_t m_consumed{0};
 };
 
+/**
+ * A mocked Sock alternative that allows providing the data to be returned by Recv()
+ * and inspecting the data that has been supplied to Send().
+ */
+class DynSock : public ZeroSock
+{
+public:
+    /**
+     * Unidirectional bytes or CNetMessage queue (FIFO).
+     */
+    class Pipe
+    {
+    public:
+        /**
+         * Get bytes and remove them from the pipe.
+         * @param[in] buf Destination to write bytes to.
+         * @param[in] len Write up to this number of bytes.
+         * @param[in] flags Same as the flags of `recv(2)`. Just `MSG_PEEK` is honored.
+         * @return The number of bytes written to `buf`. `0` if `Eof()` has been called.
+         * If no bytes are available then `-1` is returned and `errno` is set to `EAGAIN`.
+         */
+        ssize_t GetBytes(void* buf, size_t len, int flags = 0) EXCLUSIVE_LOCKS_REQUIRED(!m_mutex);
+
+        /**
+         * Deserialize a `CNetMessage` and remove it from the pipe.
+         * If not enough bytes are available then the function will wait. If parsing fails
+         * or EOF is signaled to the pipe, then `std::nullopt` is returned.
+         */
+        std::optional<CNetMessage> GetNetMsg() EXCLUSIVE_LOCKS_REQUIRED(!m_mutex);
```

Comment:
f42e4f3b3b4b3ca1945d1ea298b443f1cecaf2ea: this only works with v1 transport. In both v2 transport and the (proposed) Stratum v2 transport (#29432) bytes on the wire need further processing to reconstruct the underlying message.

For those it's more useful to have a `GetBytes(size_t n)` helper method that waits until `n` bytes have been received.

(Specifically for the Stratum v2 I'm also trying to completely avoid a dependency on `CNetMessage`, but the design for that is still in flux)

Reply:
Like this?

<details>
<summary>Add GetBytesWait()</summary>

```diff
commit 2e8674478813e2e672e5eb7dbb87f28ee11a7268
Parent: ad806a0969226806bb33a890356b89e39904c1a8
Author:     Vasil Dimov <vd@FreeBSD.org>
AuthorDate: Tue Feb 27 14:59:04 2024 +0100
Commit:     Vasil Dimov <vd@FreeBSD.org>
CommitDate: Tue Feb 27 15:48:18 2024 +0100
gpg: Signature made Tue Feb 27 15:48:19 2024 CET
gpg:                using RSA key E64D8D45614DB07545D9CCC154DF06F64B55CBBF
gpg: Good signature from "Vasil Dimov <vd@myforest.net>" [ultimate]
gpg:                 aka "Vasil Dimov <vd@FreeBSD.org>" [ultimate]
gpg:                 aka "Vasil Dimov <vasild@gmail.com>" [ultimate]


    test: add blocking "get bytes" method to DynSock

diff --git a/src/test/util/net.h b/src/test/util/net.h
index 368c84d622..d03674cb23 100644
--- a/src/test/util/net.h
+++ b/src/test/util/net.h
@@ -224,12 +224,20 @@ public:
          * @param[in] flags Same as the flags of `recv(2)`. Just `MSG_PEEK` is honored.
          * @return The number of bytes written to `buf`. `0` if `Eof()` has been called.
          * If no bytes are available then `-1` is returned and `errno` is set to `EAGAIN`.
          */
         ssize_t GetBytes(void* buf, size_t len, int flags = 0) EXCLUSIVE_LOCKS_REQUIRED(!m_mutex);
 
+        /**
+         * Wait for `len` bytes and remove them from the pipe.
+         * @param[in] buf Destination to write bytes to.
+         * @param[in] len Write this number of bytes to `buf` or less if EOF is signaled.
+         * @return The number of bytes written to `buf`. Less than `len` means EOF was signaled.
+         */
+        size_t GetBytesWait(void* buf, size_t len) EXCLUSIVE_LOCKS_REQUIRED(!m_mutex);
+
         /**
          * Deserialize a `CNetMessage` and remove it from the pipe.
          * If not enough bytes are available then the function will wait. If parsing fails
          * or EOF is signaled to the pipe, then `std::nullopt` is returned.
          */
         std::optional<CNetMessage> GetNetMsg() EXCLUSIVE_LOCKS_REQUIRED(!m_mutex);
@@ -243,13 +251,13 @@ public:
          * Construct and push CNetMessage to the pipe.
          */
         template <typename... Args>
         void PushNetMsg(const std::string& type, Args&&... payload) EXCLUSIVE_LOCKS_REQUIRED(!m_mutex);
 
         /**
-         * Signal end-of-file on the receiving end (`GetBytes()` or `GetNetMsg()`).
+         * Signal end-of-file on the receiving end (`GetBytes()`, `GetBytesWait()` or `GetNetMsg()`).
          */
         void Eof() EXCLUSIVE_LOCKS_REQUIRED(!m_mutex);
 
     private:
         /**
          * Return when there is some data to read or EOF has been signaled.
diff --git a/src/test/util/net.cpp b/src/test/util/net.cpp
index 0fee2041f0..2bd4815e89 100644
--- a/src/test/util/net.cpp
+++ b/src/test/util/net.cpp
@@ -256,12 +256,31 @@ ssize_t DynSock::Pipe::GetBytes(void* buf, size_t len, int flags)
         m_data.erase(m_data.begin(), m_data.begin() + read_bytes);
     }
 
     return read_bytes;
 }
 
+size_t DynSock::Pipe::GetBytesWait(void* buf, size_t len)
+{
+    WAIT_LOCK(m_mutex, lock);
+
+    size_t total_read_bytes{0};
+
+    do {
+        WaitForDataOrEof(lock);
+        if (!m_data.empty()) {
+            const size_t read_bytes{std::min(len - total_read_bytes, m_data.size())};
+            std::memcpy(reinterpret_cast<uint8_t*>(buf) + total_read_bytes, m_data.data(), read_bytes);
+            m_data.erase(m_data.begin(), m_data.begin() + read_bytes);
+            total_read_bytes += read_bytes;
+        }
+    } while (total_read_bytes < len && !m_eof);
+
+    return total_read_bytes;
+}
+
 std::optional<CNetMessage> DynSock::Pipe::GetNetMsg()
 {
     V1Transport transport{NodeId{0}};
 
     {
         WAIT_LOCK(m_mutex, lock);
@@ -309,12 +328,22 @@ void DynSock::Pipe::Eof()
 {
     LOCK(m_mutex);
     m_eof = true;
     m_cond.notify_all();
 }
 
+void DynSock::Pipe::WaitForDataOrEof(UniqueLock<Mutex>& lock)
+{
+    Assert(lock.mutex() == &m_mutex);
+
+    m_cond.wait(lock, [&]() EXCLUSIVE_LOCKS_REQUIRED(m_mutex) {
+        AssertLockHeld(m_mutex);
+        return !m_data.empty() || m_eof;
+    });
+}
+
 DynSock::DynSock(std::shared_ptr<Pipes> pipes) : m_pipes{pipes} {}
 
 DynSock::~DynSock()
 {
     m_pipes->send.Eof();
 }
```

I did not add that method in this PR because it would be unused. Can be added any time later as part of a PR that needs it.

Reply:
[WaitForDataOrEof.txt](https://github.com/bitcoin/bitcoin/files/14422575/WaitForDataOrEof.txt)

Reply:
Something like that yes. Maybe it can be included and used a helper function by `GetNetMsg`?

Reply:
I do not think so (or maybe I just don't see it). Inside `GetNetMsg()` we don't know how many bytes are needed. We throw something at `transport.ReceivedBytes()` and it tells us how much of that it consumed.

Reply:
`size_t len` could be an optional argument. Not sure if that otherwise creates a mess though.

Reply:
You mean to combine the two methods `GetBytes()` and `GetBytesWait()`? I considered this but ditched the idea. The bodies of the methods are too different and the `len` argument cannot be omitted as it tells the size of the buffer.